### PR TITLE
Added dftx ICXCloseOffer

### LINKED ISSUE #533: `DfTx` - ICX
> Similar to https://github.com/DeFiCh/jellyfish/issues/108, this issue tracks the implementation of custom DfTx in `jellyfish-transaction` and `jellyfish-transaction-builder`. Must implement in `jellyfish-transaction` and test in `jellyfish-transaction-builder` for each PR.

### ICX DfTX

- [x] ICXCreateOrder: `'1'` @Jouzo  #608 
- [x] ICXMakeOffer: `'2'` @Jouzo #572
- [x] ICXSubmitDFCHTLC: `'3'` @surangap https://github.com/DeFiCh/jellyfish/pull/561
- [x] ICXSubmitEXTHTLC: `'4'` @surangap https://github.com/DeFiCh/jellyfish/pull/589
- [x] ICXClaimDFCHTLC: `'5'` @surangap  https://github.com/DeFiCh/jellyfish/pull/593
- [x] ICXCloseOrder: `'6'` @Jouzo #578 
- [ ] ICXCloseOffer: `'7'` @Jouzo #579

### PR DESCRIPTION
**Pull request by Jouzo**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Adds support to create ICXCloseOffer custom transactions from jellyfish side.

#### Which issue(s) does this PR fixes?:

Fixes #533

### CODE REVIEW

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_offer.test.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    const closeOffer: ICXCloseOffer = {
+      offerTx: makeOfferTxId
+    }
+    const txn = await builder.icxorderbook.closeOffer(closeOffer, script)
+
+    const encoded: string = OP_CODES.OP_DEFI_TX_ICX_CLOSE_OFFER(closeOffer).asBuffer().toString('hex')
+    const expectedRedeemScript = `6a${encoded}`
+
+    const outs = await sendTransaction(testing.container, txn)
+    expect(outs.length).toEqual(2)
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_offer.test.ts:62]: Remove this line

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_offer.test.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    expect(outs.length).toStrictEqual(2)
+    expect(outs[0].value).toStrictEqual(0)
+    expect(outs[0].n).toStrictEqual(0)
+    expect(outs[0].tokenId).toStrictEqual(0)
+    expect(outs[0].scriptPubKey.asm.startsWith('OP_RETURN 4466547837')).toStrictEqual(true)
+    expect(outs[0].scriptPubKey.hex).toStrictEqual(expectedRedeemScript)
+    expect(outs[0].scriptPubKey.type).toStrictEqual('nulldata')
+
+    expect(outs[1].value).toEqual(expect.any(Number))
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_offer.test.ts:71]: toStrictEqual

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_order.test.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    const outs = await sendTransaction(testing.container, txn)
+    expect(outs.length).toStrictEqual(2)
+    expect(outs[0].value).toStrictEqual(0)
+    expect(outs[0].n).toStrictEqual(0)
+    expect(outs[0].tokenId).toStrictEqual(0)
+    expect(outs[0].scriptPubKey.asm.startsWith('OP_RETURN 4466547836')).toStrictEqual(true)
+    expect(outs[0].scriptPubKey.hex).toStrictEqual(expectedRedeemScript)
+    expect(outs[0].scriptPubKey.type).toStrictEqual('nulldata')
+
+    expect(outs[1].value).toEqual(expect.any(Number))
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_order.test.ts:68]: Same

File: packages/jellyfish-transaction/src/script/dftx/dftx_icxorderbook.ts [hunk 4]
```diff
@@ -78,6 +78,51 @@ export class CICXMakeOffer extends ComposableBuffer<ICXMakeOffer> {
...
+    return [
+      ComposableBuffer.hexBEBufferLE(32, () => co.orderTx, v => co.orderTx = v)
+    ]
+  }
+}
+
+/**
+ * ICXCloseOffer DeFi Transaction
+ */
+
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/src/script/dftx/dftx_icxorderbook.ts:106]: Extra break line

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    const buffer = toBuffer(stack)
+    expect(buffer.toString('hex')).toStrictEqual(hex)
+    expect((stack[1] as OP_DEFI_TX).tx.type).toStrictEqual(0x36)
+  })
+})
+
+describe('CloseOrder', () => {
+  const header = '6a254466547836' // OP_RETURN(0x6a) (length 37 = 0x25) CDfTx.SIGNATURE(0x44665478) CICXCloseOrder.OP_CODE(0x36)
+  const data = '3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152' // ICXCloseOrder.orderTx[LE](0x5211d3ef795d4129c9ebfbf1a534583692b3c1d42d0b1af79b4575e30d38b33b)
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts:42]: 5211d3ef795d4129c9ebfbf1a534583692b3c1d42d0b1af79b4575e30d38b33b should be
3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOffer.test.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    )
+    const buffer = toBuffer(stack)
+    expect(buffer.toString('hex')).toStrictEqual(hex)
+    expect((stack[1] as OP_DEFI_TX).tx.type).toStrictEqual(0x37)
+  })
+})
+
+describe('CloseOffer', () => {
+  const header = '6a254466547837' // OP_RETURN(0x6a) (length 37 = 0x25) CDfTx.SIGNATURE(0x44665478) CICXCloseOffer.OP_CODE(0x37)
+  const data = '3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152' // ICXCloseOffer.offerTx[LE](0x5211d3ef795d4129c9ebfbf1a534583692b3c1d42d0b1af79b4575e30d38b33b)
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOffer.test.ts:41]: 5211d3ef795d4129c9ebfbf1a534583692b3c1d42d0b1af79b4575e30d38b33b should be
3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_offer.test.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    expect(outs[1].scriptPubKey.addresses[0]).toStrictEqual(await providers.getAddress())
+
+    const listOrdersAfter = await testing.rpc.icxorderbook.listOrders({ orderTx })
+    const offerAfter = listOrdersAfter[makeOfferTxId]
+
+    expect(offerBefore).toBeDefined()
+    expect(offerAfter).toBeUndefined()
+  })
+
+  it('should not close ICX offer with wrong order id length', async () => {
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_icx_close_offer.test.ts:84]:
```suggestion
  it('should not close ICX offer with wrong offer id length', async () => {
```

File: packages/jellyfish-transaction/src/script/mapping.ts [hunk 8]
```diff
@@ -414,6 +434,7 @@ export const OP_CODES = {
       data: icxClaimDFCHTLC
     })
   },
+
```
Comment (surangap) [inline packages/jellyfish-transaction/src/script/mapping.ts:437]: extra line
```suggestion
```

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts (hunk 5 above)
Comment (Jouzo) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts]: The documentation or the string ? Not sure to understand

Comment (jingyi2811) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts]: Ya, documentation. If you look at other files, their comment value is a substring of **data** variable

And here 5211d3ef795d4129c9ebfbf1a534583692b3c1d42d0b1af79b4575e30d38b33b is not a substring of data variable

Comment (jingyi2811) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts]: I think this line should be replaced with

const data = '3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152' // ICXCloseOrder.orderTx[LE](0x3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152)

Comment (Jouzo) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts]: In that case, it is not a substring as there is only one property in CloseOrder interface

Comment (surangap) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts]: minor
orderTx in LE in this case should be `0x3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152`.
@Jouzo closeorder test also has it other way around.

Comment (Jouzo) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOrder.test.ts]: the orderTx is `0x5211d3ef795d4129c9ebfbf1a534583692b3c1d42d0b1af79b4575e30d38b33b` but written to buffer as LE -> `3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152` ?

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOffer.test.ts (hunk 6 above)
Comment (surangap) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_icxorderbook/CloseOffer.test.ts:41]:
```suggestion
  const data = '3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152' // ICXCloseOffer.offerTx[LE](0x3bb3380de375459bf71a0b2dd4c1b392365834a5f1fbebc929415d79efd31152)
```

**Review by jingyi2811 — APPROVED** [review]

**Review by surangap — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**